base/linalg/lq.jl

```diff
@@ -176,17 +176,41 @@ A_mul_Bc!(A::StridedMatrix{T}, B::LQPackedQ{T}) where {T<:BlasReal} =
 A_mul_Bc!(A::StridedMatrix{T}, B::LQPackedQ{T}) where {T<:BlasComplex} =
     LAPACK.ormlq!('R', 'C', B.factors, B.τ, A)
 
-# out-of-place right-application of LQPackedQs
-# unlike their in-place equivalents, these methods: (1) check whether the applied-to
-# matrix's (A's) appropriate dimension (columns for A_*, rows for Ac_*) matches the
-# number of columns (nQ) of the LQPackedQ (Q), as the underlying LAPACK routine (ormlq)
-# treats the implicit Q as its (nQ-by-nQ) square form; and (2) if the preceding dimensions
-# do not match, these methods check whether the appropriate dimension of A instead matches
-# the number of rows of the matrix of which Q is a factor (i.e. size(Q.factors, 1)),
-# and if so zero-extends A as necessary for check (1) to pass (if possible).
-*(A::StridedVecOrMat, Q::LQPackedQ) = _A_mul_Bq(A_mul_B!, A, Q)
-A_mul_Bc(A::StridedVecOrMat, Q::LQPackedQ) = _A_mul_Bq(A_mul_Bc!, A, Q)
-function _A_mul_Bq(A_mul_Bop!::FT, A::StridedVecOrMat, Q::LQPackedQ) where FT<:Function
+# out-of-place right application of LQPackedQs
+#
+# LQPackedQ's out-of-place multiplication behavior is context dependent. specifically,
+# if the inner dimension in the multiplication is the LQPackedQ's second dimension,
+# the LQPackedQ behaves like its full/square form. if the inner dimension in the
+# multiplication is the LQPackedQ's first dimension, the LQPackedQ behaves like either
+# its full/square form or its thin form depending on the shape of the other object
+# involved in the multiplication. we treat these cases separately.
+#
+# (1) the inner dimension in the multiplication is the LQPackedQ's second dimension.
+# in this case, the LQPackedQ behaves like its square/full form.
+#
+function A_mul_Bc(A::StridedVecOrMat, Q::LQPackedQ)
+    TR = promote_type(eltype(A), eltype(Q))
+    return A_mul_Bc!(copy_oftype(A, TR), convert(AbstractMatrix{TR}, Q))
+end
+function Ac_mul_Bc(A::StridedMatrix, Q::LQPackedQ)
+    TR = promote_type(eltype(A), eltype(Q))
+    C = adjoint!(similar(A, TR, reverse(size(A))), A)
+    return A_mul_Bc!(C, convert(AbstractMatrix{TR}, Q))
+end
+#
+# (2) the inner dimension in the multiplication is the LQPackedQ's first dimension.
+# in this case, the LQPackedQ behaves like either its square/full form or its
+# thin form depending on the shape of the other object in the multiplication.
```

Comment:
Maybe it should be pointed out that *thin* actually means *wide* in the case of LQ`.

Reply:
Why?!? Can it be renamed?

Reply:
Agreed. Though the "thin" decomposition/factorization terminology inherited from QR seems the most widespread, I have also seen "reduced" and "truncated" decomposition/factorization, both of which seem like better choices when applied to orthogonal decompositions/factorizations generally (and particularly "truncated" for its descriptiveness). Perhaps we should select one of the latter terms (and an equivalent term for the "full" or "complete" decomposition/factorization) and use that term consistently for all orthogonal decompositions/factorizations? If so, perhaps opening a separate issue for that change would be best? Best!

Reply:
> Why?!? Can it be renamed?

The `LQ` is simply `QR` transposed so when `Q` is thin in `QR` it is actually wide in `LQ`. There are two cases here: the comment and the keyword. Here I just thought of extending the comment to point out that the matrix is wide. For the keyword, notice that you only use when materializing a dense `Q` so it is not really relevant when you just use the factorization. However, it is a bit unfortunate that `thin` means `wide` in `lq`. Maybe we should negate the argument and have a `square` keyword instead.

Reply:
Considerations for future work in any case? :)

Reply:
Is this a question of tall & thin vs short & wide? I.e. The transpose would be thin.

Reply:
If I understand your question, then yes: In a thin(/truncated/reduced) QR factorization of a tall/thin matrix, `Q` is tall/thin and `R` square. Similarly, in a "thin" LQ factorization of the (short/wide) transpose of that tall/thin matrix, `L` is square and `Q` is short/wide (the thin LQ factorization of the transposed matrix being the transpose of the thin QR factorization of the original matrix). That such LQ factorizations are sometimes called thin seems to be an artifact of the literature's emphasis on QR / history. Similar considerations apply to RQ and QL factorizations. Hence the more general terms truncated and reduced might be better. Best!

Reply:
So maybe just clearly. Document that thin means tall and skinny or short and wide?

Reply:
> Considerations for future work in any case? :)

Yes. I'm just suggesting extending the comment a bit. Maybe also the docstring. Changes to the keyword argument should be a separate PR.

Reply:
I extended the docstring with a note briefly describing the somewhat confusing "thin"/"tall", "wide"/"short", "truncated"/"reduced" situation, and revised the language in the code comments to use the hopefully less confusing "truncated" and "square". Thoughts? Thanks!

Reply:
Should we just use "truncated" consistently?

Reply:
https://github.com/JuliaLang/julia/pull/23813#discussion_r140558814 :). That change sounds like work for a future pull request though. Best!

Reply:
I agree. We should consider changing the language here but in a separate PR.

Reply:
Tracking issue: https://github.com/JuliaLang/julia/issues/23882